plugins/janus_nosip.c

```diff
@@ -0,0 +1,2213 @@
+
+static void janus_nosip_message_free(janus_nosip_message *msg) {
+	if(!msg || msg == &exit_message)
+		return;
+
+	msg->handle = NULL;
+
+	g_free(msg->transaction);
+	msg->transaction = NULL;
+	if(msg->message)
+		json_decref(msg->message);
+	msg->message = NULL;
+	if(msg->jsep)
+		json_decref(msg->jsep);
+	msg->jsep = NULL;
+
+	g_free(msg);
+}
+
+
+typedef struct janus_nosip_rtp_context {
+	/* Needed to fix seq and ts in case RTP sources change */
+	uint32_t a_last_ssrc, a_last_ts, a_base_ts, a_base_ts_prev,
+			v_last_ssrc, v_last_ts, v_base_ts, v_base_ts_prev;
+	uint16_t a_last_seq, a_base_seq, a_base_seq_prev,
+			v_last_seq, v_base_seq, v_base_seq_prev;
+} janus_nosip_rtp_context;
+
+typedef struct janus_nosip_media {
+	char *remote_ip;
+	int ready:1;
+	gboolean autoack;
+	gboolean require_srtp, has_srtp_local, has_srtp_remote;
+	int has_audio:1;
+	int audio_rtp_fd, audio_rtcp_fd;
+	int local_audio_rtp_port, remote_audio_rtp_port;
+	int local_audio_rtcp_port, remote_audio_rtcp_port;
+	guint32 audio_ssrc, audio_ssrc_peer;
+	int audio_pt;
+	const char *audio_pt_name;
+	srtp_t audio_srtp_in, audio_srtp_out;
+	srtp_policy_t audio_remote_policy, audio_local_policy;
+	int audio_srtp_suite_in, audio_srtp_suite_out;
+	gboolean audio_send;
+	int has_video:1;
+	int video_rtp_fd, video_rtcp_fd;
+	int local_video_rtp_port, remote_video_rtp_port;
+	int local_video_rtcp_port, remote_video_rtcp_port;
+	guint32 video_ssrc, video_ssrc_peer;
+	int video_pt;
+	const char *video_pt_name;
+	srtp_t video_srtp_in, video_srtp_out;
+	srtp_policy_t video_remote_policy, video_local_policy;
+	int video_srtp_suite_in, video_srtp_suite_out;
+	gboolean video_send;
+	janus_nosip_rtp_context context;
+	int pipefd[2];
+	gboolean updated;
+} janus_nosip_media;
+
+typedef struct janus_nosip_session {
+	janus_plugin_session *handle;
+	char *info;					/* Opaque identifier that the application can set */
+	janus_nosip_media media;	/* Media gatewaying stuff (same stuff as the SIP plugin) */
+	janus_sdp *sdp;				/* The SDP this user sent */
+	janus_recorder *arc;		/* The Janus recorder instance for this user's audio, if enabled */
+	janus_recorder *arc_peer;	/* The Janus recorder instance for the peer's audio, if enabled */
+	janus_recorder *vrc;		/* The Janus recorder instance for this user's video, if enabled */
+	janus_recorder *vrc_peer;	/* The Janus recorder instance for the peer's video, if enabled */
+	janus_mutex rec_mutex;		/* Mutex to protect the recorders from race conditions */
+	volatile gint hangingup;
+	gint64 destroyed;	/* Time at which this session was marked as destroyed */
+	janus_mutex mutex;
+} janus_nosip_session;
+static GHashTable *sessions;
+static GList *old_sessions;
+static janus_mutex sessions_mutex;
+
+
+/* SRTP stuff (in case we need SDES) */
+#define SRTP_MASTER_KEY_LENGTH	16
+#define SRTP_MASTER_SALT_LENGTH	14
+#define SRTP_MASTER_LENGTH (SRTP_MASTER_KEY_LENGTH + SRTP_MASTER_SALT_LENGTH)
+static const char *janus_nosip_srtp_error[] =
+{
```

Comment:
this is also present in the SIP plugin, right? Maybe some srtputils.c module is in order?